fml/shared_thread_merger_impl.h

```diff
@@ -0,0 +1,64 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef FLUTTER_FML_SHARED_THREAD_MERGER_H_
+#define FLUTTER_FML_SHARED_THREAD_MERGER_H_
+
+#include <condition_variable>
+#include <mutex>
+
+#include "flutter/fml/macros.h"
+#include "flutter/fml/memory/ref_counted.h"
+#include "flutter/fml/message_loop_task_queues.h"
+
+namespace fml {
+
+class MessageLoopImpl;
+class RasterThreadMerger;
+
+struct ThreadMergerKey {
+  TaskQueueId owner;
+  TaskQueueId subsumed;
+  bool operator<(const ThreadMergerKey& other) const {
+    if (owner == other.owner) {
+      return subsumed < other.subsumed;
+    } else {
+      return owner < other.owner;
+    }
+  }
+};
+
+class SharedThreadMergerImpl
+    : public fml::RefCountedThreadSafe<SharedThreadMergerImpl> {
+ public:
+  SharedThreadMergerImpl(fml::TaskQueueId owner, fml::TaskQueueId subsumed);
+  static fml::RefPtr<SharedThreadMergerImpl> GetSharedImpl(
+      fml::TaskQueueId owner,
+      fml::TaskQueueId subsumed);
+  void RecordMergerCaller(RasterThreadMerger* caller);
+  bool MergeWithLease(size_t lease_term);
+  /// TODO DOC
+  bool UnMergeNowIfLastOne(RasterThreadMerger* caller);
+  void ExtendLeaseTo(size_t lease_term);
+  bool IsMergedUnSafe() const;
+  bool DecrementLease();
+
+ private:
+  static const int kLeaseNotSet;
+  fml::TaskQueueId owner_;
+  fml::TaskQueueId subsumed_;
+  fml::RefPtr<fml::MessageLoopTaskQueues> task_queues_;
+  std::atomic_int lease_term_;
+  std::mutex mutex_;
+  std::set<fml::RasterThreadMerger*> merge_callers_;
```

Comment:
I think this instance variable might merit a docstring.  It's a bit odd since it has raw pointers.

Reply:
we could also get away with storing the task queue id of the caller, rather than a reference to the thread merger itself.

Reply:
@iskakaushik When using the flutter 2.0 lightweight engines groups, the two callers of rasterizer are both from same raster_queue_id, the engines shares the same raster queue.  Only when using the classic standalone engines, the queue id of the caller are different. 😄

Reply:
Maybe a typedef'd `void*` would be better since it would discourage anyone from trying to use them for anything other than identity checks?   `typedef void* RasterThreadMergerId`?

Reply:
Good suggestion, let me do the modifications